src/sundials/sundials_context.c

```diff
@@ -18,82 +18,157 @@
+
+  /* SUNContext_Create cannot assert or log since the SUNContext is not yet
+   * created */
+  if (!sunctx) {
+    return SUN_ERR_MALLOC_FAIL;
+  }
 
+  SUNFunctionBegin(sunctx);
+  
 #ifdef SUNDIALS_ADIAK_ENABLED 
   adiak_init(&comm);
   sunAdiakCollectMetadata();
 #endif
 
+  do {
 #if SUNDIALS_LOGGING_LEVEL > 0 
 #if SUNDIALS_MPI_ENABLED
-  if (SUNLogger_CreateFromEnv(comm, &logger))
+    err = SUNLogger_CreateFromEnv(comm, &logger); 
+    SUNCheckCallNoRet(err); if (err) { break; }
 #else
-  if (SUNLogger_CreateFromEnv(SUN_COMM_NULL, &logger))
+    err = SUNLogger_CreateFromEnv(SUN_COMM_NULL, &logger); 
+    SUNCheckCallNoRet(err); if (err) { break; }
 #endif
-  {
-    return (-1);
-  }
 #else
-#if SUNDIALS_MPI_ENABLED
-  if (SUNLogger_Create(comm, 0, &logger))
-#else
-  if (SUNLogger_Create(SUN_COMM_NULL, 0, &logger))
+    err = SUNLogger_Create(SUN_COMM_NULL, 0, &logger); 
+    SUNCheckCallNoRet(err); if (err) { break; }
+    err = SUNLogger_SetErrorFilename(logger, ""); 
+    SUNCheckCallNoRet(err); if (err) { break; }
+    err = SUNLogger_SetWarningFilename(logger, "");
+    SUNCheckCallNoRet(err); if (err) { break; }
+    err = SUNLogger_SetInfoFilename(logger, "");
+    SUNCheckCallNoRet(err); if (err) { break; }
+    err = SUNLogger_SetDebugFilename(logger, "");
+    SUNCheckCallNoRet(err); if (err) { break; }
 #endif
-  {
-    return (-1);
-  }
-  SUNLogger_SetErrorFilename(logger, "");
-  SUNLogger_SetWarningFilename(logger, "");
-  SUNLogger_SetInfoFilename(logger, "");
-  SUNLogger_SetDebugFilename(logger, "");
+
+#if defined(SUNDIALS_BUILD_WITH_PROFILING) && !defined(SUNDIALS_CALIPER_ENABLED)
+    err = SUNProfiler_Create(comm, "SUNContext Default", &profiler); 
+    SUNCheckCallNoRet(err); if (err) { break; }
 #endif
 
-  *sunctx = NULL;
-  *sunctx = (SUNContext)malloc(sizeof(struct _SUNContext));
+    err = SUNErrHandler_Create(SUNLogErrHandlerFn, NULL, &eh);
+    SUNCheckCallNoRet(err); if (err) { break; }
 
-  if (*sunctx == NULL)
-  {
+    sunctx->logger        = logger;
+    sunctx->own_logger    = logger != NULL;
+    sunctx->profiler      = profiler;
+    sunctx->own_profiler  = profiler != NULL;
+    sunctx->last_err      = 0;
+    sunctx->err_handler   = eh;
+    sunctx->comm          = comm;
+  } while (0);
+
+  if (err) {
 #if defined(SUNDIALS_BUILD_WITH_PROFILING) && !defined(SUNDIALS_CALIPER_ENABLED)
-    SUNProfiler_Free(&profiler);
+    SUNCheckCallNoRet(SUNProfiler_Free(&profiler));
 #endif
-    SUNLogger_Destroy(&logger);
-    return (-1);
+    SUNCheckCallNoRet(SUNLogger_Destroy(&logger));
+    free(sunctx);
+  } else {
+    *sunctx_out = sunctx;
   }
 
-  (*sunctx)->logger       = logger;
-  (*sunctx)->own_logger   = logger != NULL;
-  (*sunctx)->profiler     = profiler;
-  (*sunctx)->own_profiler = profiler != NULL;
+  return err;
+}
+
+SUNErrCode SUNContext_GetLastError(SUNContext sunctx)
+{
+  if (!sunctx) { return SUN_ERR_SUNCTX_CORRUPT; }
 
-  return (0);
+  SUNFunctionBegin(sunctx);
```

Comment:
I think it makes sense to follow the rule since it won't do any harm and if we ever use `SUNFunctionBegin` for additional purposes it will already be there.